src/main/java/net/coderbot/iris/pipeline/TransformPatcher.java

```diff
@@ -0,0 +1,512 @@
+		protected abstract String getMainContent();
+
+		{
+			detectionResult("irisMain");
+			wrapTarget("main");
+			injectionLocation(InjectionPoint.BEFORE_EOF);
+			injectionExternalDeclarations(CachePolicy.ON_JOB);
+		}
+
+		@Override
+		protected Collection<String> getInjectionExternalDeclarations() {
+			return CompatUtil.listOf("void main() { " + getMainContent() + " }");
+		}
+	}
+
+	static {
+		/**
+		 * PREV TODO: Only do the NewLines patches if the source code isn't from
+		 * gbuffers_lines
+		 */
+
+		TokenFilter<Parameters> parseTokenFilter = new ChannelFilter<Parameters>(TokenChannel.PREPROCESSOR) {
+			@Override
+			public boolean isTokenAllowed(Token token) {
+				if (!super.isTokenAllowed(token)) {
+					throw new SemanticException("Unparsed preprocessor directives such as '" + token.getText()
+							+ "' may not be present at this stage of shader processing!");
+				}
+				return true;
+			}
+		};
+
+		// setup the transformations and even loose phases if necessary
+		LifecycleUser<Parameters> detectReserved = new SearchTerminals<Parameters>()
+				.singleTarget(
+						new ThrowTargetImpl<Parameters>(
+								"iris_",
+								"Detected a potential reference to unstable and internal Iris shader interfaces (iris_). This isn't currently supported."))
+				.requireFullMatch(false);
+
+		// #region patchAttributes
+		LifecycleUser<Parameters> replaceEntityColorDeclaration = new Transformation<Parameters>() {
+			@Override
+			protected void setupGraph() {
+				addEndDependent(new WalkPhase<Parameters>() {
+					ParseTreePattern entityColorPattern;
+
+					@Override
+					public void init() {
+						entityColorPattern = compilePattern(
+								"uniform vec4 entityColor;",
+								GLSLParser.RULE_externalDeclaration);
+					}
+
+					@Override
+					public void enterExternalDeclaration(ExternalDeclarationContext ctx) {
+						ParseTreeMatch match = entityColorPattern.match(ctx);
+						if (match.succeeded()) {
+							removeNode(ctx);
+						}
+					}
+				});
+
+				if (getJobParameters().type == ShaderType.GEOMETRY) {
+					chainDependent(new SearchTerminals<Parameters>().singleTarget(
+							new ParsedReplaceTargetImpl<>("entityColor", "entityColor[0]", GLSLParser::expression)));
+				}
+
+				chainDependent(new RunPhase<Parameters>() {
+					@Override
+					protected void run(TranslationUnitContext ctx) {
+						switch (getJobParameters().type) {
+							case VERTEX:
+								injectExternalDeclarations(InjectionPoint.BEFORE_DECLARATIONS,
+										"uniform sampler2D iris_overlay;",
+										"varying vec4 entityColor;");
+								break;
+							case GEOMETRY:
+								injectExternalDeclarations(InjectionPoint.BEFORE_DECLARATIONS,
+										"out vec4 entityColorGS;",
+										"in vec4 entityColor[];");
+								break;
+							case FRAGMENT:
+								injectExternalDeclaration(InjectionPoint.BEFORE_DECLARATIONS, "varying vec4 entityColor;");
+								break;
+						}
+					}
+				});
+			}
+		};
```

Comment:
This should probably be abstracted since turning a uniform into a vertex attribute that gets shuffled through the patches is going to be a somewhat common thing.

Reply:
This transformation could have pretty many variations in an abstract form:
- The pattern
- The (conditional) parsed replacement input and output
- If something has to be detected with `ThrowTarget`
- Which injections happen on what shader types and how many there are (could be made variable if necessary)

Which of these would be shared or variable in an abstracted version of this? If too much is abstracted away, it'll become cumbersome to use the abstracted transformation with varying (or dynamic, i.e. based on job parameters) requirements. (see the config system in `WrapIdentifier`)